Helm: Add tsdb schema config

faustodavid opened the pull request:
**What this PR does / why we need it**:
To take advantage of the new tsdb schema.

I have also updated the `query_scheduler.max_outstanding_requests_per_tenant` and `querier. max_concurrent ` following the recommendations from this post:
https://grafana.com/docs/loki/latest/operations/storage/tsdb

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [ ] Tests updated
- [x] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`
- [x] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

MichelHollands commented:
@faustodavid Thanks for this contribution. This will have an impact on people upgrading an existing Helm installation. Can you make the date for the tsdb_index configurable so it can be in the future? The migration impact should also be documented. Given the big change this could be 5.9 or even a 6.0 release.

faustodavid commented:
> @faustodavid Thanks for this contribution. This will have an impact on people upgrading an existing Helm installation. Can you make the date for the tsdb_index configurable so it can be in the future? The migration impact should also be documented. Given the big change this could be 5.9 or even a 6.0 release.

I moved the configuration to `schemaConfig` and added a comment with a link to the TSDB documentation.

Should I add something more to explain the impact?

MichelHollands commented:
@faustodavid Can you add **BREAKING** to the CHANGELOG entry. It should also be 6.0 as this is a breaking change. Also please add instruction on how to upgrade (ie set the TSDB start date in the future). Did you test this upgrade?

faustodavid commented:
> @faustodavid Can you add **BREAKING** to the CHANGELOG entry. It should also be 6.0 as this is a breaking change. Also please add instruction on how to upgrade (ie set the TSDB start date in the future). Did you test this upgrade?
> 
> Here's an example of a new version in the changelog with breaking entries: https://github.com/grafana/loki/blob/main/production/helm/loki/CHANGELOG.md#40

Hey @MichelHollands, I ran a test by deploying the helm version 5.8.6 to AKS and then applied the changes from this PR. All went well and no Loki error logs popped out except when the pods were restarting.

Once the updated version was rolled out, the TSDB index got added and Loki could read both schemas with ease.

<img width="407" alt="image" src="https://github.com/grafana/loki/assets/2423149/2cbccb3b-3001-458a-a26b-9644e3c1155f">

I sent out 5 messages when the `boltdb` schema was turned on and then sent 5 more when we switched to `tsdb`.

<img width="1120" alt="image" src="https://github.com/grafana/loki/assets/2423149/f19429be-ec11-4ba6-8ba1-e75ec1daa6e9">

Here's what I've figured: the date in the TSDB schema doesn't need to be future-dated. It just needs to come after the boltdb date. So, if folks aren't ready to use the new schema right away, they can just set this date to the future.

sandeepsukhani commented:
> > @faustodavid Can you add **BREAKING** to the CHANGELOG entry. It should also be 6.0 as this is a breaking change. Also please add instruction on how to upgrade (ie set the TSDB start date in the future). Did you test this upgrade?
> > Here's an example of a new version in the changelog with breaking entries: https://github.com/grafana/loki/blob/main/production/helm/loki/CHANGELOG.md#40
> 
> Hey @MichelHollands, I ran a test by deploying the helm version 5.8.6 to AKS and then applied the changes from this PR. All went well and no Loki error logs popped out except when the pods were restarting.
> 
> Once the updated version was rolled out, the TSDB index got added and Loki could read both schemas with ease.
> 
> <img alt="image" width="407" src="https://user-images.githubusercontent.com/2423149/251579669-2cbccb3b-3001-458a-a26b-9644e3c1155f.png">
> I sent out 5 messages when the `boltdb` schema was turned on and then sent 5 more when we switched to `tsdb`.
> 
> <img alt="image" width="1120" src="https://user-images.githubusercontent.com/2423149/251579980-f19429be-ec11-4ba6-8ba1-e75ec1daa6e9.png">
> Here's what I've figured: the date in the TSDB schema doesn't need to be future-dated. It just needs to come after the boltdb date. So, if folks aren't ready to use the new schema right away, they can just set this date to the future.

I do not think it would work because we use schema config at query time to find the relevant tables. You need to check if the logs are really coming from both tables or just the tsdb. You can verify that by renaming(adding any letter at the beginning) or removing either of the tables in the object storage and running the queries again after restarting Loki.

faustodavid commented:
> > > @faustodavid Can you add **BREAKING** to the CHANGELOG entry. It should also be 6.0 as this is a breaking change. Also please add instruction on how to upgrade (ie set the TSDB start date in the future). Did you test this upgrade?
> > > Here's an example of a new version in the changelog with breaking entries: https://github.com/grafana/loki/blob/main/production/helm/loki/CHANGELOG.md#40
> > 
> > 
> > Hey @MichelHollands, I ran a test by deploying the helm version 5.8.6 to AKS and then applied the changes from this PR. All went well and no Loki error logs popped out except when the pods were restarting.
> > Once the updated version was rolled out, the TSDB index got added and Loki could read both schemas with ease.
> > <img alt="image" width="407" src="https://user-images.githubusercontent.com/2423149/251579669-2cbccb3b-3001-458a-a26b-9644e3c1155f.png">
> > I sent out 5 messages when the `boltdb` schema was turned on and then sent 5 more when we switched to `tsdb`.
> > <img alt="image" width="1120" src="https://user-images.githubusercontent.com/2423149/251579980-f19429be-ec11-4ba6-8ba1-e75ec1daa6e9.png">
> > Here's what I've figured: the date in the TSDB schema doesn't need to be future-dated. It just needs to come after the boltdb date. So, if folks aren't ready to use the new schema right away, they can just set this date to the future.
> 
> I do not think it would work because we use schema config at query time to find the relevant tables. You need to check if the logs are really coming from both tables or just the tsdb. You can verify that by renaming(adding any letter at the beginning) or removing either of the tables in the object storage and running the queries again after restarting Loki.

@sandeepsukhani, I did a few tests and everything worked smoothly. When I deleted the `boltdb` index folder and restarted the pods, I couldn't find the logs I previously made in `boltdb`. I only found those I logged with `tsdb`. But, after putting the index folder back and restarting, the logs were visible again.

faustodavid commented:
I have also tried locally using the example https://github.com/grafana/loki/tree/main/examples/getting-started.

I did the following steps:
* Run the docker-compose
* Wait some minutes to have some sample data
* Do POST on the writer `http://localhost:3102/ingester/shutdown`
* Shutdown the reader
* Add tsdb schema config with a date older than today.
* Turn on the writer pod, then the reader
* Query the reader and validate that old data is there.
* Just to double check Do POST on the writer `http://localhost:3102/ingester/shutdown`
* Validate that both index folders are in Minio
* Turn on the writer and reader and test the data again

MichelHollands commented:
@faustodavid I'm afraid I can't approve this as it is. Adding the TSDB there with a hardcoded start date will mess things up when people upgrade.  Note that https://grafana.com/docs/loki/latest/operations/storage/tsdb/ says the TSDB date has to be in the future.

However if you remove the TSDB part from the schemaConfig setting then it can be merged. The other settings are useful and the having the schemaConfig configurable will help with the migration to TSDB. The TSDB part can be mentioned in the CHANGELOG as something that can be used for new users or added with a start date in the future for existing users.
